transit_odp/pipelines/pipelines/dataset_etl/transform.py

```diff
@@ -201,3 +352,100 @@ def get_most_common_districts(stops: pd.DataFrame):
+                ["atco_code", "flexible_location"]
+            ]
+            filtered_stop_points_naptan_id.set_index("atco_code", inplace=True)
+            return filtered_stop_points_naptan_id
+        return pd.DataFrame()
+
+    def create_flexible_timing_link(self, flexible_journey_details, flexible_jp_to_jps):
+        def get_stop_sequence(group, jp_section_ref):
+            group["order"] = range(len(group))
+            group["route_link_ref"] = (
+                str(jp_section_ref) + "RL" + (group["order"] + 1).astype(str)
+            )
+            group["jp_timing_link_id"] = (
+                str(jp_section_ref) + "TL" + (group["order"] + 1).astype(str)
+            )
+            group["from_stop_ref"] = group["atco_code"].shift(0)
+            group["to_stop_ref"] = group["atco_code"].shift(-1)
+            return group.dropna(subset=["to_stop_ref"])
+
+        if not flexible_journey_details.empty and not flexible_jp_to_jps.empty:
+            flexible_jp_to_jps = flexible_jp_to_jps.reset_index()[
+                ["journey_pattern_id", "jp_section_ref"]
+            ]
+            journey_patterns = flexible_journey_details.reset_index().merge(
+                flexible_jp_to_jps,
+                how="left",
+                left_on="journey_pattern_id",
+                right_on="journey_pattern_id",
+            )
+            grouped = journey_patterns.groupby("jp_section_ref")
+            flexible_timing_links = pd.concat(
+                [
+                    get_stop_sequence(group, jp_section_ref)
+                    for jp_section_ref, group in grouped
+                ]
+            )
+            flexible_timing_links = (
+                flexible_timing_links[
+                    [
+                        "file_id",
+                        "jp_timing_link_id",
+                        "route_link_ref",
+                        "jp_section_ref",
+                        "order",
+                        "from_stop_ref",
+                        "to_stop_ref",
+                    ]
+                ]
+                .rename(columns={"jp_section_ref": "jp_section_id"})
+                .set_index(["file_id", "jp_section_id"])
+            )
+            return flexible_timing_links
+        return pd.DataFrame()
+
+    def create_flexible_jps(self, flexible_journey_details):
+        flexible_jps = flexible_journey_details.reset_index()[
+            ["file_id", "journey_pattern_id", "service_code"]
+        ].drop_duplicates(["file_id", "journey_pattern_id"])
+        flexible_jps["jp_section_ref"] = flexible_jps.apply(
+            lambda x: hash(
+                str(x.journey_pattern_id) + str(x.service_code) + str(x.file_id)
+            ),
+            axis=1,
+        )
+        flexible_jp_to_jps = flexible_jps[
+            ["file_id", "journey_pattern_id", "jp_section_ref"]
+        ].drop_duplicates(["jp_section_ref"])
```

Comment:
Under which scenario will there be duplicate `jp_section_ref`?

Reply:
drop_duplicates for jp_section_ref can be removed 
and now i am generating uuid for each jp_section_ref not relying on the python hash function